modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4ChunkedContinuationsIT.java

```diff
@@ -0,0 +1,657 @@
+                            computingContinuation = false;
+                            return getChunkedBody();
+                        }));
+                    }
+
+                    @Override
+                    public ReleasableBytesReference encodeChunk(int sizeHint, Recycler<BytesRef> recycler) {
+                        assertTrue(lines.hasNext());
+                        refs.mustIncRef();
+                        return new ReleasableBytesReference(new BytesArray(lines.next()), refs::decRef);
+                    }
+
+                    @Override
+                    public String getResponseContentTypeString() {
+                        return TEXT_CONTENT_TYPE;
+                    }
+                };
+            }
+        }
+
+        public static class TransportInfiniteContinuationsAction extends TransportAction<Request, Response> {
+            private final ExecutorService executor;
+
+            @Inject
+            public TransportInfiniteContinuationsAction(ActionFilters actionFilters, TransportService transportService) {
+                super(TYPE.name(), actionFilters, transportService.getTaskManager());
+                this.executor = transportService.getThreadPool().executor(ThreadPool.Names.GENERIC);
+            }
+
+            @Override
+            protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+                executor.execute(
+                    ActionRunnable.supply(ActionTestUtils.assertNoFailureListener(listener::onResponse), () -> new Response(executor))
+                );
+            }
+        }
+
+        @Override
+        public Collection<RestHandler> getRestHandlers(
+            Settings settings,
+            NamedWriteableRegistry namedWriteableRegistry,
+            RestController restController,
+            ClusterSettings clusterSettings,
+            IndexScopedSettings indexScopedSettings,
+            SettingsFilter settingsFilter,
+            IndexNameExpressionResolver indexNameExpressionResolver,
+            Supplier<DiscoveryNodes> nodesInCluster,
+            Predicate<NodeFeature> clusterSupportsFeature
+        ) {
+            return List.of(new BaseRestHandler() {
+                @Override
+                public String getName() {
+                    return ROUTE;
+                }
+
+                @Override
+                public List<Route> routes() {
+                    return List.of(new Route(GET, ROUTE));
+                }
+
+                @Override
+                protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) {
+                    final var localRefs = refs; // single volatile read
+                    if (localRefs != null && localRefs.tryIncRef()) {
+                        return new RestChannelConsumer() {
+                            @Override
+                            public void close() {
+                                refs.decRef();
```

Comment:
Nit: maybe `localRefs.decRef()` here and below?

Reply:
It's important that `refs` hasn't been changed here - modified this to use `localRefs` but added an explicit assertion about that.